modules/core/src/main/java/org/apache/ignite/internal/processors/platform/cache/query/PlatformContinuousQueryImpl.java

```diff
@@ -169,12 +174,21 @@ private static CacheEntryEventFilter getJavaFilter(Object filter, GridKernalCont
         }
         finally {
             lock.writeLock().unlock();
+            startLatch.countDown();
         }
     }
 
     /** {@inheritDoc} */
     @Override public void onUpdated(Iterable evts) throws CacheEntryListenerException {
-        lock.readLock().lock();
+        // First, use startLatch to ensure full initialization.
+        try {
+            startLatch.await();
+        } catch (InterruptedException e) {
+            throw new IgniteInterruptedException(e);
+        }
+
+        if (!lock.readLock().tryLock())
+            return; // Query is being closed (see tryLock reasoning in #evaluate).
```

Comment:
It seems to me, we should throw a new exception here. 
`throw new CacheEntryListenerException("Failed to notify listener because it has been closed.");`

Reply:
I don't think it makes sense to throw an exception from `onUpdate` when the query is closed. The user has stopped the continuous query, which means he is not interested in updates anymore. We can just ignore them.

I've replaced existing `throw` in `onUpdate` with `return` as well, to be consistent.